### PR TITLE
Add `arraySymmetricDifference` function

### LINKED ISSUE #61673: Implement `arraySymmetricDifference` function
**Use case**

Would be nice addition to `arrayIntersect` or `arrayUnion` https://github.com/ClickHouse/ClickHouse/issues/61553

**Additional context**

ClickHouse already has `arrayDifference`, but this it does a completely other thing.

### PR DESCRIPTION
**Pull request by pheepa**
Exported logic from `arrayIntersect` to `arrayLogicalFunction` class that covers handling both intersection and symmetric difference. Added `arraySymmetricDifference` function with docs and tests.

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added function `arraySymmetricDifference`. This closes: https://github.com/ClickHouse/ClickHouse/issues/61673

### Documentation entry for user-facing changes

- [x] Documentation is written

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 4b6b13aa5814d9189c4908f40d67a945cb94c977 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62262/4b6b13aa5814d9189c4908f40d67a945cb94c977/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>❌ failure</td></tr>
<tbody>
</table>

File: docs/en/sql-reference/functions/array-functions.md [hunk 1]
```diff
@@ -1434,6 +1434,24 @@ SELECT
...
+## arraySymmetricDifference(arr)
+
+Takes multiple arrays, returns an array with elements that are presented not in all source arrays.
+
+Example:
+
+``` sql
+SELECT
+    arrayIntersect([1, 2], [1, 2], [1, 2]) AS no_symmetric_difference
+    arrayIntersect([1, 2], [1, 2], [1, 3]) AS symmetric_difference,
```
Comment (nikitamikhaylov) [inline docs/en/sql-reference/functions/array-functions.md:1446]: It uses `arrayIntersect`.

**Comment by nikitamikhaylov** [comment]
The extracted part should be also useful for https://github.com/ClickHouse/ClickHouse/issues/61553

File: src/Functions/array/arrayLogicalFunction.cpp [hunk 2]
```diff
@@ -0,0 +1,548 @@
+#include <Columns/ColumnArray.h>
```
Comment (rschu1ze) [inline src/Functions/array/arrayLogicalFunction.cpp:1]: Comment related to this file in general. Please add in strategic places high-level comments what the code/algorithms is doing. For a casual reader, it is almost impossible to follow without tedious debugging to understand what is happening in this file.

(as an example, you could look at src/Functions/array/arrayFold.cpp which I commented recently after a severe bug was found in the initial implementation)

File: src/Functions/array/arraySymmetricDifference.cpp [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+{
+
+class FunctionArraySymmetricDifference : public FunctionArrayLogical
+{
+public:
+    static constexpr auto name = "arraySymmetricDifference";
+
+    static FunctionPtr create(ContextPtr) { return std::make_shared<FunctionArraySymmetricDifference>(); }
+
+    FunctionArraySymmetricDifference() : FunctionArrayLogical(false, name) { }
```
[the remark this answers is not in the document]
Comment (pheepa) [reply src/Functions/array/arraySymmetricDifference.cpp:17]: i will move bool intersect to template, thx
is it still okay to have this FunctionArrayLogical(name)?

[the remark this answers is not in the document]
Comment (rschu1ze) [reply src/Functions/array/arraySymmetricDifference.cpp:17]: I don't think we need to pass it into the ctor as a parameter.

We could do sth like

```
template <bool intersect>
class FunctionArrayLogicalBase
{
    void executeImpl(...)
    {
        if constexpr (intersect)
        {
            ...
        }

}

[the remark this answers is not in the document]
Comment (pheepa) [reply src/Functions/array/arraySymmetricDifference.cpp:17]: > FunctionArrayLogical::getReturnTypeImpl) can no longer be in a source file, they need to be in a header

executeImpl should be in header as well, doesn't it?

[the remark this answers is not in the document]
Comment (rschu1ze) [reply src/Functions/array/arraySymmetricDifference.cpp:17]: Yes, with templates, file "src/Functions/array/arrayLogicalFunction.cpp" will no longer exist (i.e. everything will be in the header)

File: tests/queries/0_stateless/03033_array_symmetric_difference.sql [hunk 4]
```diff
@@ -0,0 +1,14 @@
+select arraySymmetricDifference([], []);
```
[the remark this answers is not in the document]
Comment (pheepa) [reply tests/queries/0_stateless/03033_array_symmetric_difference.sql]: added some complex test

not so sure how to test getReturnTypeImpl, some hints here?

**Comment by pheepa** [comment]
Can anyone explain me why my fast test is failing?

I see `[Errno 111] Connection refused` in logs

File: tests/queries/0_stateless/03033_array_symmetric_difference.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (rschu1ze) [reply tests/queries/0_stateless/03033_array_symmetric_difference.sql]: Thanks. Based on the existing tests, I think that `getReturnTypeImpl` is already tested fairly well. No need to add more tests.

**Comment by rschu1ze** [comment]
> Can anyone explain me why my fast test is failing?
> 
> I see `[Errno 111] Connection refused` in logs It does not look like my fault...

It was due to shaky infrastructure. But I also see some test failures which are fixed in upstream. Kindly merge from `master` to include the last source state.

**Comment by rschu1ze** [comment]
Stupid question. Currently, we get this:

```sql
SELECT arraySymmetricDifference([1, 2], [1, 2], [1, 3]);
-- returns [3]
```

Given that the function is documented as `Takes multiple arrays, returns an array with elements which are not present in all source arrays.`, should that not return `[2, 3]`?

**Comment by pheepa** [comment]
> Stupid question. Currently, we get this:
> 
> ```sql
> SELECT arraySymmetricDifference([1, 2], [1, 2], [1, 3]);
> -- returns [3]
> ```
> 
> Given that the function is documented as `Takes multiple arrays, returns an array with elements which are not present in all source arrays.`, should that not return `[2, 3]`?

yeah, it should, thank you 
stupid it is, first pr into ch, im learning

**Comment by rschu1ze** [comment]
In any case: thank you for working on this!

**Comment by pheepa** [comment]
Hi @rschu1ze 
I see you have been working on https://github.com/ClickHouse/ClickHouse/issues/57816 in the past, and now after refactoring of arrayIntersect i got simillar issue
But it appears even with arrayConcat, mentioned it here 
https://github.com/ClickHouse/ClickHouse/pull/57879#issuecomment-2046539263

**Comment by rschu1ze** [comment]
@pheepa Upon a quick look, this PR looks good, I only updated some expected results.

What I would love to see are more code comments in arrayLogicalFunctionBase.h. This may feel tedious but it will be very helpful for anyone working on the code in future. Perhaps you like to add more of them? Thanks 100x.

**Comment by pheepa** [comment]
I believe that comments would help to everybody as well
The main reason why I have not added so much is because part with return types is hard to read for me as well, I will figure out and add as much as I could

Anyway this pr cant be merged until I resolve problem with casting. I see author cased all columns before function process them, but now I see that it is the main problem why tests array_fold and array_jaccard_index are failing
Give me another couple of days :)

thx for your support and being nice

**Comment by pheepa** [comment]
Still no luck here, I am sure I have problems with providing context 
On line 140 in arrayLogicalFunctionBase context field is used but it's not initialized so segfault happens, can't finish this task without hints here
Tell me please if I am missing something

**Comment by rschu1ze** [comment]
@pheepa Is there a certain SELECT statement which crashes?

**Comment by pheepa** [comment]
yes, `select [toUInt32(1)] as arr1, [toUInt64(1)] as arr2, arrayIntersect(arr1, arr2)`

**Comment by pheepa** [comment]
> yes, `select [toUInt32(1)] as arr1, [toUInt64(1)] as arr2, arrayIntersect(arr1, arr2)`

fixed this btw

**Comment by rschu1ze** [comment]
The test failures are related, could you please check them?

**Comment by pheepa** [comment]
> Also, the commit history of this PR is really messy.
I completely agree, lets drop this one, I will reopen new with fixes


> The test failures are related, could you please check them?
I am trying to fix them

**Pull request closed without merging**